<!DOCTYPE HTML>
<html>
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8"></meta>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"></meta>

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"></link>
  	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.0/css/bootstrap.min.css"></link>
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css"></link>
    
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  	<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>


</head>

<body>
	<nav style="margin:0px" class="navbar navbar-expand-lg navbar-dark fixed-top bg-dark">
  		<a class="navbar-brand" th:href="@{/}" href="frontpage.html">Variant3D</a>
		<button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarsExampleDefault" aria-controls="navbarsExampleDefault" aria-expanded="false" aria-label="Toggle navigation">
    		<span class="navbar-toggler-icon"></span>
  		</button>
  		
  		<div class="collapse navbar-collapse" id="navbarsExampleDefault">
    		<ul class="navbar-nav mr-auto">
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/pageapi}" href="api.html">API</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/download}" href="download.html">Download</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/clients}" href="clients.html">Clients</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/statistics}" href="statistics.html">Statistics</a>
      			</li>     			
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" target="_blank" href="https://github.com/juexinwang/G2Smutation">Github</a>
      			</li> 
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/tutorial}" href="tutorial.html">Tutorial</a>
      			</li> 
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/about}" href="about.html">About</a>
      			</li>    			
    		</ul>
  		</div>
	</nav>
	
	<div class="jumbotron">
        <section id="header" class="jumbotron text-center">
        	<h2 class="display-3" style="margin-top: 0.6em">Variant3D (Variants on 3D Structure)</h2>
        	<p class="lead">An auto-updating database collecting and annotating protein point mutations on 3D structures</p>
        	<div class="row justify-content-md-center">
        	<div class="col-xl-5">
        	<input type="text" class="form-control" placeholder="Search variants in a gene or protein e.g. P53_HUMAN" aria-label="Recipient's username"></input>
        	e.g. P53_HUMAN or P04637 or ENSP00000269305.4 or ENSG00000141510.11
        	<button type="submit" class="btn btn-default">Submit</button>
        	</div></div>
    	</section>
    </div>
	
	<div class="container">
	<div class="row">
        <div class="col-md-8 blog-main">
        	<div class="blog-post">
	            <p style="text-align:justify;font-size:1.2em">
	            Variant3D is an automatically weekly updated database collecting and annotating all human-related point residue variants on 3D protein structures. Built on genomic to structure mapping tool <a target="_blank" href="https://g2s.genomenexus.org">G2S</a>, Variant3D systematically collects all point mutations mapping from human UniProt/Ensemble proteins in PDB. Aligning back to genomic locations, these mutations on 3D structures are annotated by structural annotation via analyzing PDB, variant annotation as DBSNP, genetics annotation as COSMIC, disease annotation as Clinvar/TCGA. Variant3D provides a user-friendly API with REST-inspired design, and it can be used by various clients and other bioinformatics tools for bringing 3D structures into the genomic variant analysis. 
				</p>
				<hr></hr>
			</div>
			
			<div class="blog-post">
				<h4 class="blog-post-title">Highlights:</h4>
				<ol style="list-style-type: none;padding:0px;font-size:1.2em">
	            	<li>• Collecting all human point mutations on 3D structure in the whole PDB.</li>
	            	<li>• Structure annotation by 2nd Structure, protein surface location, and sequence/structrue domain</li>
	            	<li>• 3D protein variants annotated by mapping dbSNP/COSMIC/TCGA on human genome</li>
					<li>• Automatically update every week after <a target="_blank" href="http://www.wwpdb.org/download/downloads#rcsbpdb">PDB updates</a></li>
					<li>• RESTful <a target="_blank" th:href="@{/pageapi}" href="api.html">API</a> for users and <a target="_blank" th:href="@{/clients}" href="clients.html">clients</a></li>
	            </ol>
				<hr></hr>
			</div>
        	

			<!--         	
          	<div class="blog-post">
            	<h4 class="blog-post-title">Citation:</h4>
          		<p style="text-align:justify">
				G2S: A web-service for annotating genomic variants on 3D protein
				structures, Juexin Wang, Robert Sheridan, S Onur Sumer, Nikolaus
				Schultz, Dong Xu, Jianjiong Gao, Bioinformatics, 34(11), 1949-1950, <a target="_blank" href="https://doi.org/10.1093/bioinformatics/bty047">(link)</a>
				</p>
				<hr></hr>
			</div>
			!-->
			
			
        </div>

        <aside class="col-md-4 blog-sidebar">			
			<div class="p-3 mb-3 bg-light rounded">
            	<h4>Contact</h4>
            	<p style="text-align:justify">
            	If you have any questions and feedback, please contact:</p>
            	<p>Juexin Wang (wangjue(at)missouri.edu)</p><p>Jianjiong Gao (jgao@cbio.mskcc.org).
            	</p>
          	</div>			
        </aside>
      </div>
      
      <div>
		<footer>
		<p>© 2019 Variant3D</p>
		</footer>
	  </div>
	  </div>
	
</body>
</html>
