<!DOCTYPE HTML>
<html>
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8"></meta>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"></meta>

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"></link>
  	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.0/css/bootstrap.min.css"></link>
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css"></link>
    
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  	<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>


</head>

<body>
	<nav style="margin:0px" class="navbar navbar-expand-lg navbar-dark fixed-top bg-dark">
  		<a class="navbar-brand" th:href="@{/}" href="frontpage.html">Variant3D</a>
		<button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarsExampleDefault" aria-controls="navbarsExampleDefault" aria-expanded="false" aria-label="Toggle navigation">
    		<span class="navbar-toggler-icon"></span>
  		</button>
  		
  		<div class="collapse navbar-collapse" id="navbarsExampleDefault">
    		<ul class="navbar-nav mr-auto">
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/pageapi}" href="api.html">API</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/sequence}" href="sequence.html">Sequence</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/database}" href="database.html">Database</a>
      			</li>
      			<li class="nav-item">
        			<a style="font-size:1.2em" class="nav-link" th:href="@{/statistics}" href="statistics.html">Statistics</a>
      			</li>
    		</ul>
  		</div>
	</nav>
	
	<div class="jumbotron">
        <section id="header" class="jumbotron text-center">
        	<h2 class="display-3" style="margin-top: 0.6em">Variant3D (Variants on 3D Structure)</h2>
        	<p class="lead">An auto-updating database annotating genomic/protein variants on 3D structures</p>
        	<div class="col-md-4">
        	<input type="text" class="form-control" placeholder="Search for a gene or protein e.g. P53_HUMAN" aria-label="Recipient's username"></input>
        	</div>
        	e.g. P53_HUMAN
    	</section>
    </div>
	
	<div class="container">
	<div class="row">
        <div class="col-md-8 blog-main">
        	<div class="blog-post">
            	<h2 class="blog-post-title">Why variant3D?</h2>
	            <p style="text-align:justify;font-size:1.2em">
				There is a great interest and need of mapping and analyzing germline and somatic genomic variants generated by recent large-scale
				sequencing efforts in the context of 3D protein structures. Several tools have been developed to address the need, including
				<a target="_blank" href="https://www.ebi.ac.uk/pdbe/docs/sifts/">SIFTS</a>, a service by <a target="_blank" href="http://www.rcsb.org/pdb/chromosome.do">PDB</a>,
				and <a target="_blank" href="http://www.sheba-cancer.org.il/G23D">G23D</a>. However, none of them provides a web service that support automated 
				analysis programmatically. We have developed a comprehensive web service that support automated mapping and annotating genomic 
				variants in 3D protein structures,
				</p>
				<hr></hr>
			</div>
			
        	<div class="blog-post">
	            <h2 class="blog-post-title">Two components</h2>
	            <p style="text-align:justify;font-size:1.2em;margin:0px"><b>Variant3D API</b> (<a target="_blank" th:href="@{/pageapi}" href="api.html">link</a>):Query by Uniprot/Human Ensembl/Genomic Position.</p>
	            <p style="text-align:justify;font-size:1.2em"><b>Sequence Web Interface</b> (<a target="_blank" th:href="@{/sequence}" href="sequence.html">link</a>):Query by any protein sequences.</p>
	            <hr></hr>
        	</div>
        	
        	<div class="blog-post"> 
	            <h2 class="blog-post-title">G2S Supports</h2>
	            <ol style="list-style-type: none;padding:0px;font-size:1.2em">
	            	<li>• Retrieving aligned protein structure chains for an Ensembl/UniProt entry.</li>
					<li>• Retrieving aligned protein structure chains for any protein sequence.</li>
					<li>• Retrieving residue mapping between protein sequences and protein structures.</li>
					<li>• Retrieving structural position of a genomic variant.</li>
	            </ol>
	            <p style="text-align:justify;font-size:1.2em">
				The web service will return high confidence pre-calculated mapping of
				the protein/residue aligned to the protein structures from the PDB in
				the JSON format. A fixed e-value score threshold are predefined with
				several alignment quality metrics to enable further filtering if
				desired. Our system makes frequent (weekly) updates to the set of
				pre-computed alignments as new pdb structures are added to RCSB,
				while keeping the set of protein sequences from public protein
				sequence databases fixed. Less frequently, the system is
				re-initialized using an updated set of protein sequences, and during
				re-initialization an all-against-all alignment process is required.
				This web service was initially designed and implemented for mapping
				and visualizing cancer mutations in the <a target="_blank" href="http://www.cbioportal.org/">cBioPortal</a>, 
				a leading resource for studying cancer genomics.
				The cBioPortal has been used by thousands of researchers (currently
				there are in average about 3,000 visit daily), mostly individuals
				outside of the authors’ group. Recently, we separated and generalized
				the web service from the cBioPortal and established the G2S web
				service in the project Genome Nexus for annotating genomic variants.
				The web service will continue serving the cBioPortal as well as any
				other applications. The codes of the web service and website are
				open-source and available at <a href="https://github.com/cBioPortal/pdb-annotation">GitHub</a>.
				</p>
            	<hr></hr>
        	</div>

          	<div class="blog-post">
            	<h2 class="blog-post-title">History</h2>
            	<p class="blog-post-meta">Latest update in February 12, 2019</p>
	            <ul style="list-style-type: none;padding:0px;font-size:1.2em">
	              <li >2019-02-12 - Fix a bug of sequence id mixmatch in residue mapping.</li>
	              <li>2018-12-19 - Fix a bug of alignment mismatch in residue mapping.</li>
	              <li>2018-10-29 - Fix minor bug API 500 error when upstream API return error.</li>
	              <li>2018-10-05 - Add support of ENST/ENSG mapping in all versions.</li>
	              <li>2018-10-04 - Add support of mapping dbSNP ID to PDB.</li>
	              <li>2018-06-30 - Fix small bug of mapping hgvs-grch37 genes.</li>
	              <li>2017-11-30 - Initial Version released.</li>
	            </ul>
	            <hr></hr>
          	</div>
          	
          	<div class="blog-post">
            	<h4 class="blog-post-title">Citation:</h4>
          		<p style="text-align:justify">
				G2S: A web-service for annotating genomic variants on 3D protein
				structures, Juexin Wang, Robert Sheridan, S Onur Sumer, Nikolaus
				Schultz, Dong Xu, Jianjiong Gao, Bioinformatics, 34(11), 1949-1950, <a target="_blank" href="https://doi.org/10.1093/bioinformatics/bty047">(link)</a>
				</p>
				<hr></hr>
			</div>
			
			<div>
			<footer>
				<p>© 2017 G2S Web Service</p>
			</footer>
			</div>
        </div>

        <aside class="col-md-4 blog-sidebar">
        	<div class="p-3 mb-3 bg-light rounded">
            	<h4>About</h4>
            	<p style="text-align:justify">
            	The goal of the G2S is to offer a high-accuracy, high-reliable, auto-updating computational services providing 
            	protein sequence to tertiary structure alignments in residue level. <a target="_blank" th:href="@{/about}" href="about.html">(Learn More)</a>
            	</p>
          	</div>
			
			<div class="p-3 mb-3 bg-light rounded">
            	<h4>Contact</h4>
            	<p style="text-align:justify">
            	If you have any questions and feedback, please contact:Juexin Wang (wangjue(at)missouri.edu) and Jianjiong Gao (jgao@cbio.mskcc.org).
            	</p>
          	</div>
			
          	<div class="p-3">
            	<h4>Elsewhere</h4>
            	<ol class="list-unstyled">
              		<li><a target="_blank" href="https://github.com/cBioPortal/pdb-annotation">GitHub</a></li>
              		<li><a target="_blank" th:href="@{/clients}" href="clients.html">Clients</a></li>
              		<li><a target="_blank" th:href="@{/statistics}" href="statistics.html">Statistics</a></li>
            	</ol>
          	</div>
        </aside>
      </div>
	  </div>
	
</body>
</html>
